src/hotspot/share/oops/arrayKlass.cpp

```diff
@@ -125,7 +125,7 @@ GrowableArray<Klass*>* ArrayKlass::compute_secondary_supers(int num_extra_slots,
 }
 
 objArrayOop ArrayKlass::allocate_arrayArray(int n, int length, TRAPS) {
-  check_array_allocation_length(length, arrayOopDesc::max_array_length(T_ARRAY), CHECK_NULL);
+  check_array_allocation_length(length, arrayOopDesc::max_array_length(element_type()), CHECK_NULL);
```

Comment:
Wait, why? This method allocates the array of the "this"-typed arrays, I think. I.e. if `this` is `int[]`, then `element_type()` is `int`, and then `allocate_arrayArray(y, len)` allocates `int[length][]`, and we still need to check `length` against `max_array_length(T_ARRAY)`? Which is also why we do `objArrayOopDesc::object_size` at the very next line.

Reply:
Right. My bad.